chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,637 @@
+    let increase_pool_balance_undo = accounting_adapter
+        .operations(TransactionSource::Chain(block_id))
+        .increase_pool_pledge_amount(pool_id, total_owner_reward)?;
+
+    let undos = delegation_undos
+        .into_iter()
+        .chain(vec![increase_pool_balance_undo].into_iter())
+        .collect();
+
+    Ok(AccountingBlockRewardUndo::new(undos))
+}
+
+fn calculate_pool_owner_reward(total_reward: Amount, pool_data: &PoolData) -> Option<Amount> {
+    (total_reward - pool_data.cost_per_block())
+        .and_then(|v| v * pool_data.margin_ratio_per_thousand().value().into())
+        .and_then(|v| v / 1000)
+        .and_then(|v| v + pool_data.cost_per_block())
+}
+
+/// The reward is distributed among delegations proportionally to their balance
+fn distribute_delegations_pos_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    delegation_shares: &BTreeMap<DelegationId, Amount>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_delegations_balance: Amount,
+    total_delegations_reward: Amount,
+) -> Result<(Vec<PoSAccountingUndo>, Amount), ConnectTransactionError> {
+    let rewards_per_delegation = calculate_rewards_per_delegation(
+        delegation_shares,
+        pool_id,
+        total_delegations_balance,
+        total_delegations_reward,
+    )?;
+
+    // increase the delegation balances
+    let delegation_undos = rewards_per_delegation
+        .iter()
+        .map(|(delegation_id, reward)| {
+            accounting_adapter
+                .operations(TransactionSource::Chain(block_id))
+                .delegate_staking(*delegation_id, *reward)
+                .map_err(ConnectTransactionError::PoSAccountingError)
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Due to integer arithmetics there can be a small remainder after all the delegations distributed.
+    // This remainder goes to the pool's owner
+    let total_delegations_reward_distributed =
+        rewards_per_delegation.iter().try_fold(Amount::ZERO, |acc, (_, v)| {
+            (acc + *v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                block_id, pool_id,
+            ))
+        })?;
+
+    let delegations_reward_remainder =
+        (total_delegations_reward - total_delegations_reward_distributed).ok_or(
+            ConnectTransactionError::DistributedDelegationsRewardExceedTotal(
+                pool_id,
+                block_id,
+                total_delegations_reward_distributed,
+                total_delegations_reward,
+            ),
+        )?;
+    Ok((delegation_undos, delegations_reward_remainder))
+}
+
+fn calculate_rewards_per_delegation(
+    delegation_shares: &BTreeMap<DelegationId, Amount>,
+    pool_id: PoolId,
+    total_delegations_amount: Amount,
+    total_delegations_reward_amount: Amount,
+) -> Result<Vec<(DelegationId, Amount)>, ConnectTransactionError> {
+    assert_eq_size!(Amount, Uint128);
```

Comment:
Why Uint128 not u128?

Also please add a comment explaining what this assert is valuable here. Something like "this condition is necessary to ensure that the multiplication of balance and rewards won't overflow; if this is to change, please ensure that the output of the operations below has twice as many bits".

Though I would really prefer to have a runtime check for this... but apparently Uint256 is infected with C/C++ issues.